source/Halibut/Transport/Protocol/MessageExchangeStream.cs

```diff
@@ -246,10 +246,20 @@ void ExpectServerIdentity()
 
         T ReadBsonMessage<T>()
         {
+            //TODO: Remove these debugging checks once we're done investigating https://trello.com/c/0gHf4v7m/2974-null-ref-exceptions-in-halibut
+            if (stream == null)
+                throw new Exception("stream is null");
+            if (serializer == null)
+                throw new Exception("serializer is null");
+            
             using (var zip = new DeflateStream(stream, CompressionMode.Decompress, true))
             using (var bson = new BsonDataReader(zip) { CloseInput = false })
             {
-                return (T)serializer.Deserialize<MessageEnvelope>(bson).Message;
+                var messageEnvelope = serializer.Deserialize<MessageEnvelope>(bson);
+                if (messageEnvelope == null)
```

Comment:
I'm guessing it's more likely that the envelope is null? If that is the case, any ideas what could cause the stream to be in that kind of state?